### PR TITLE
[Process] Allow inheriting env vars instead of replacing them

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #18796 |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

File: src/Symfony/Component/Process/Process.php [hunk 1]
```diff
@@ -290,10 +291,21 @@ public function start(callable $callback = null)
...
+            $env = null;
+        } else {
+            $env = $this->env;
+        }
 
-        $this->process = proc_open($commandline, $descriptors, $this->processPipes->pipes, $this->cwd, $this->env, $this->options);
+        $this->process = @proc_open($commandline, $descriptors, $this->processPipes->pipes, $this->cwd, $env, $this->options);
 
         if (!is_resource($this->process)) {
+            echo $commandline;
```
Comment (stof) [inline src/Symfony/Component/Process/Process.php]: should be removed

File: src/Symfony/Component/Process/ProcessBuilder.php [hunk 2]
```diff
@@ -275,6 +275,9 @@ public function getProcess()
 
         $process = new Process($script, $this->cwd, $env, $this->input, $this->timeout, $options);
 
+        if ($this->inheritEnv) {
```
Comment (stof) [inline src/Symfony/Component/Process/ProcessBuilder.php]: you should probably remove the old merging of the environment then

**Comment by nicolas-grekas** [comment]
Comments addressed, all green :)

File: src/Symfony/Component/Process/Process.php [hunk 3]
```diff
@@ -267,9 +268,22 @@ public function start(callable $callback = null)
...
         $commandline = $this->commandline;
+        $envline = '';
 
+        if (null !== $this->env && $this->inheritEnv) {
+            if ('\\' === DIRECTORY_SEPARATOR && !empty($this->options['bypass_shell']) && !$this->enhanceWindowsCompatibility) {
+                throw new LogicException('The "bypass_shell" option must be false to inherit environment variables while enhanced Windows compatibility is off');
+            }
+            $env = '\\' === DIRECTORY_SEPARATOR ? '(SET %s)&&' : 'export %s;';
+            foreach ($this->env as $k => $v) {
+                $envline .= sprintf($env, ProcessUtils::escapeArgument("$k=$v"));
```
Comment (xabbuh) [inline src/Symfony/Component/Process/Process.php]: Don't we need one space after a value if it is followed by another key?

File: src/Symfony/Component/Process/Process.php [hunk 4]
```diff
@@ -1198,6 +1214,30 @@ public function setEnhanceSigchildCompatibility($enhance)
...
+
+        return $this;
+    }
+
+    /**
+     * Returns whether environment variables will be inherited or not.
+     *
+     * @return bool
+     */
+    public function getInheritEnvironmentVariables()
```
Comment (romainneutron) [inline src/Symfony/Component/Process/Process.php]: this sounds like a getter for Inherited env vars. What about `areEnvironmentVariablesInherited`?

Comment (nicolas-grekas) [inline src/Symfony/Component/Process/Process.php]: less discoverable maybe?

File: src/Symfony/Component/Process/Process.php [hunk 5]
```diff
@@ -1198,6 +1214,30 @@ public function setEnhanceSigchildCompatibility($enhance)
     }
 
     /**
+     * Sets whether environment variables will be inherited or not.
+     *
+     * @param bool $inheritEnv
+     *
+     * @return ProcessBuilder
+     */
+    public function inheritEnvironmentVariables($inheritEnv = true)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Process/Process.php]: I'd personally prefer `inheritEnv()` here, but I borrowed this name from ProcessBuilder, consistency first :)

File: src/Symfony/Component/Process/Process.php (hunk 4 above)
Comment (xabbuh) [inline src/Symfony/Component/Process/Process.php]: I like @romainneutron's suggestion.

Comment (nicolas-grekas) [inline src/Symfony/Component/Process/Process.php]: let's go, updated

File: src/Symfony/Component/Process/Process.php [hunk 6]
```diff
@@ -1198,6 +1214,30 @@ public function setEnhanceSigchildCompatibility($enhance)
     }
 
     /**
+     * Sets whether environment variables will be inherited or not.
+     *
+     * @param bool $inheritEnv
+     *
+     * @return ProcessBuilder
```
Comment (fabpot) [inline src/Symfony/Component/Process/Process.php]: @return is wrong

**Comment by fabpot** [comment]
👍

**Comment by stof** [comment]
:+1:

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Pull request merged**